<!DOCTYPE html><html lang="zh-CN"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1"><meta name="format-detection" content="telephone=no"><meta name="apple-mobile-web-app-capable" content="yes"><meta name="apple-mobile-web-app-status-bar-style" content="black"><link rel="icon" href="/images/icons/favicon-16x16.png?v=2.6.2" type="image/png" sizes="16x16"><link rel="icon" href="/images/icons/favicon-32x32.png?v=2.6.2" type="image/png" sizes="32x32"><meta name="description" content="细胞太多怎么办？没用的扔了就是了！">
<meta property="og:type" content="article">
<meta property="og:title" content="单细胞下采样">
<meta property="og:url" content="https://biobrick.github.io/2023/03/02/%E5%8D%95%E7%BB%86%E8%83%9E%E4%B8%8B%E9%87%87%E6%A0%B7/">
<meta property="og:site_name" content="Bio Harbor">
<meta property="og:description" content="细胞太多怎么办？没用的扔了就是了！">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2023-03-02T05:55:01.000Z">
<meta property="article:modified_time" content="2023-03-11T04:20:42.584Z">
<meta property="article:author" content="Chong Yin">
<meta name="twitter:card" content="summary"><title>单细胞下采样 | Bio Harbor</title><link ref="canonical" href="https://biobrick.github.io/2023/03/02/%E5%8D%95%E7%BB%86%E8%83%9E%E4%B8%8B%E9%87%87%E6%A0%B7/"><link rel="dns-prefetch" href="https://cdn.jsdelivr.net"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.12.1/css/all.min.css" type="text/css"><link rel="stylesheet" href="/css/index.css?v=2.6.2"><script>var Stun = window.Stun || {};
var CONFIG = {
  root: '/',
  algolia: undefined,
  assistSearch: undefined,
  fontIcon: {"prompt":{"success":"fas fa-check-circle","info":"fas fa-arrow-circle-right","warning":"fas fa-exclamation-circle","error":"fas fa-times-circle"},"copyBtn":"fas fa-copy"},
  sidebar: {"offsetTop":"20px","tocMaxDepth":6},
  header: {"enable":true,"showOnPost":true,"scrollDownIcon":false},
  postWidget: {"endText":true},
  nightMode: {"enable":true},
  back2top: {"enable":true},
  codeblock: {"style":"default","highlight":"light","wordWrap":false},
  reward: true,
  fancybox: false,
  zoomImage: {"gapAside":"20px"},
  galleryWaterfall: undefined,
  lazyload: false,
  pjax: undefined,
  externalLink: {"icon":{"enable":true,"name":"fas fa-external-link-alt"}},
  shortcuts: undefined,
  prompt: {"copyButton":"复制","copySuccess":"复制成功","copyError":"复制失败"},
  sourcePath: {"js":"js","css":"css","images":"images"},
};

window.CONFIG = CONFIG;</script><meta name="generator" content="Hexo 6.0.0"></head><body><div class="container" id="container"><header class="header" id="header"><div class="header-inner"><nav class="header-nav header-nav--fixed"><div class="header-nav-inner"><div class="header-nav-menubtn"><i class="fas fa-bars"></i></div><div class="header-nav-menu"><div class="header-nav-menu-item"><a class="header-nav-menu-item__link" href="/"><span class="header-nav-menu-item__icon"><i class="fas fa-home"></i></span><span class="header-nav-menu-item__text">首页</span></a></div><div class="header-nav-menu-item"><a class="header-nav-menu-item__link" href="/archives/"><span class="header-nav-menu-item__icon"><i class="fas fa-folder-open"></i></span><span class="header-nav-menu-item__text">归档</span></a></div></div><div class="header-nav-mode"><div class="mode"><div class="mode-track"><span class="mode-track-moon"></span><span class="mode-track-sun"></span></div><div class="mode-thumb"></div></div></div></div></nav><div class="header-banner"><div class="header-banner-info"><div class="header-banner-info__title">Bio Harbor</div><div class="header-banner-info__subtitle"></div></div></div></div></header><main class="main" id="main"><div class="main-inner"><div class="content-wrap" id="content-wrap"><div class="content" id="content"><!-- Just used to judge whether it is an article page--><div id="is-post"></div><div class="post"><header class="post-header"><h1 class="post-title">单细胞下采样</h1><div class="post-meta"><span class="post-meta-item post-meta-item--createtime"><span class="post-meta-item__icon"><i class="far fa-calendar-plus"></i></span><span class="post-meta-item__info">发表于</span><span class="post-meta-item__value">2023-03-02</span></span><span class="post-meta-item post-meta-item--updatetime"><span class="post-meta-item__icon"><i class="far fa-calendar-check"></i></span><span class="post-meta-item__info">更新于</span><span class="post-meta-item__value">2023-03-11</span></span></div></header><div class="post-body"><p>最近做了一些个超大数据集的分析，其中出现了超级大量的问题….除了数据不整洁，找不到直接适用的meta还得自己生成之外，最大的问题，就是数据实在太大了没法跑….有些分析甚至直接能给200G的服务器内存给爆了，总不能什么时候都氪金吧？而且部分方法直接不支持超大量数据或者无法并行的问题。这时候，就需要用到下采样了，理论上如果我们认可自己在细胞聚类和注释这一步中，得到的cluster的纯度，那么，取部分细胞代表整体是很合理的，当然这一步可以用ROGUE工具进行检查，下采样时，除了subset里的downsample参数，和pseudocell聚合间接下采样，或者直接分层抽样，bootstrap抽样等方法之外，还发现了一个专门的工具可以进行下采样：<span class="exturl"><a class="exturl__link" target="_blank" rel="noopener" href="https://github.com/brianhie/geosketch">geosketch</a><span class="exturl__icon"><i class="fas fa-external-link-alt"></i></span></span></p>
<p>这是个python包，但是通过reticulate可以很方便地调用且直接兼容R中的矩阵直接下采样，因而可以在纯R环境下运行：</p>
<figure class="highlight r"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#在R中需要调用reticulate使用python环境，py中则可以原生运行</span></span><br><span class="line">library(reticulate)</span><br><span class="line">library(rsvd)</span><br><span class="line"></span><br><span class="line">geosketch &lt;- import(<span class="string">&#x27;geosketch&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Generate some random data.</span></span><br><span class="line"><span class="comment">#samples in rows, features in columns</span></span><br><span class="line">X &lt;- replicate(<span class="number">20</span>, rnorm(<span class="number">1000</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment"># Get top PCs from randomized SVD.</span></span><br><span class="line">s &lt;- rsvd(X, k=<span class="number">10</span>)</span><br><span class="line">X.pcs &lt;- s$u %*% diag(s$d)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Sketch 10% of data.</span></span><br><span class="line">sketch.size &lt;- <span class="built_in">as.integer</span>(<span class="number">100</span>)</span><br><span class="line">sketch.indices &lt;- geosketch$gs(X.pcs, sketch.size, one_indexed = <span class="literal">TRUE</span>)</span><br><span class="line">print(sketch.indices)</span><br><span class="line"></span><br><span class="line"><span class="comment">#下采样：</span></span><br><span class="line">X_sub &lt;- X[unlist(sketch.indices),]</span><br><span class="line"></span><br><span class="line"><span class="comment">#有点坑啊....试试pbmc3k：</span></span><br><span class="line">library(Seurat)</span><br><span class="line">library(SeuratData)</span><br><span class="line"></span><br><span class="line">sce &lt;- pbmc3k.final</span><br><span class="line"></span><br><span class="line">DimPlot(sce)</span><br><span class="line"></span><br><span class="line">X &lt;- t(as.matrix(GetAssayData(sce,assay = <span class="string">&quot;RNA&quot;</span>,slot = <span class="string">&quot;counts&quot;</span>)))</span><br><span class="line"></span><br><span class="line">s &lt;- rsvd(X, k=<span class="number">10</span>)</span><br><span class="line">X.pcs &lt;- s$u %*% diag(s$d)</span><br><span class="line"></span><br><span class="line">sketch.size &lt;- <span class="built_in">as.integer</span>(<span class="number">500</span>)</span><br><span class="line">sketch.indices &lt;- geosketch$gs(X.pcs, sketch.size, one_indexed = <span class="literal">TRUE</span>)</span><br><span class="line">print(sketch.indices)</span><br><span class="line"></span><br><span class="line"><span class="comment">#这里技巧是，不需要获取下采样的矩阵，直接获取下采样出来的细胞名就行了</span></span><br><span class="line">X_sub_cell &lt;- rownames(X)[unlist(sketch.indices)]</span><br><span class="line">sce$BC &lt;- rownames(sce@meta.data)</span><br><span class="line"></span><br><span class="line">sce.sub &lt;- subset(sce, subset = BC %in% X_sub_cell)</span><br><span class="line"></span><br><span class="line">DimPlot(sce) + DimPlot(sce.sub)</span><br><span class="line"></span><br><span class="line">table(sce$seurat_annotations)</span><br><span class="line">table(sce.sub$seurat_annotations)  </span><br></pre></td></tr></table></div></figure>
<p>从结果上讲，还真是有点东西！即便对于最少的细胞类型，抽样上也并没有忽略掉…而且在全程分析中并没有使用过CellType这个信息，且没有缺少某种类型，证明这个方法的确比较稳健。经过反复多次尝试，确定基本上这个方法能够照顾到极不均衡数据，不至于出现一下采样直接缺少某种细胞类型的问题。<br>当然，这么方便的工具不会全无问题，关键就在其中的获取转置矩阵那一步，因为这个工具需要行是细胞，列是基因，还不支持稀疏矩阵，所以消耗上在R里有时候仍然会有问题……但这个包原生是为py开发的，python中，特别是scanpy中，表达矩阵的形式直接就是符合要求的….因而不需要任何转置操作，而且内存消耗极小，看上去python环境刻不容缓啊….</p>
<p>那么，不依托于python，有没有其他的工具呢？</p>
<p>也有，那就是metacell类工具，当然，由于metacell坑爹的开发组，这个工具没法在win平台安装，linux和unix平台都能安装，但metacell的接口写的很晦涩，结果展示做的也不怎么样，因此自然就有大佬感觉不爽，去开发新的工具，那就是<span class="exturl"><a class="exturl__link" target="_blank" rel="noopener" href="https://doi.org/10.1186/s12859-022-04861-1">Supercell</a><span class="exturl__icon"><i class="fas fa-external-link-alt"></i></span></span>，当然严格来说这类工具做的是cell aggregating，但是能够实现和下采样类似的效果，极大减少细胞数量，降低计算开销，以及提升某些分析的准确度：</p>
<figure class="highlight r"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br></pre></td><td class="code"><pre><span class="line">library(SuperCell)</span><br><span class="line"><span class="comment">#这里直接用pbmc的数据做一次单数据集尝试</span></span><br><span class="line">rm(<span class="built_in">list</span> = ls())</span><br><span class="line"></span><br><span class="line">library(SeuratData)</span><br><span class="line">library(Seurat)</span><br><span class="line"></span><br><span class="line">sce &lt;- pbmc3k.final</span><br><span class="line">sce &lt;- UpdateSeuratObject(sce)</span><br><span class="line">sce$seurat_annotations</span><br><span class="line">DimPlot(sce,reduction = <span class="string">&quot;umap&quot;</span>,group.by = <span class="string">&quot;seurat_annotations&quot;</span>)</span><br><span class="line"></span><br><span class="line">gamma &lt;- 10 <span class="comment"># Graining level，聚合的力度</span></span><br><span class="line">hvg &lt;- VariableFeatures(sce,selection.method = <span class="string">&quot;vst&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Compute metacells using SuperCell package</span></span><br><span class="line">MC &lt;- SCimplify(</span><br><span class="line">    X = GetAssayData(sce), <span class="comment"># single-cell log-normalized gene expression data</span></span><br><span class="line">    genes.use = hvg, </span><br><span class="line">    <span class="built_in">gamma</span> = <span class="built_in">gamma</span>,</span><br><span class="line">    n.pc = <span class="number">10</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Compute gene expression of metacells by simply averaging gene expression within each metacell</span></span><br><span class="line">MC.ge &lt;- supercell_GE(</span><br><span class="line">    ge = GetAssayData(sce),</span><br><span class="line">    groups = MC$membership</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">MC$cell_line &lt;- supercell_assign(</span><br><span class="line">    cluster = sce$seurat_annotations,          <span class="comment"># single-cell assignment to cell lines </span></span><br><span class="line">    supercell_membership = MC$membership,  <span class="comment"># single-cell assignment to metacells</span></span><br><span class="line">    method = <span class="string">&quot;jaccard&quot;</span> <span class="comment"># available methods are c(&quot;jaccard&quot;, &quot;relative&quot;, &quot;absolute&quot;), function&#x27;s help() for explanation</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">method_purity &lt;- <span class="built_in">c</span>(<span class="string">&quot;max_proportion&quot;</span>, <span class="string">&quot;entropy&quot;</span>)[<span class="number">1</span>]</span><br><span class="line">MC$purity &lt;- supercell_purity(</span><br><span class="line">    clusters = sce$seurat_annotations,</span><br><span class="line">    supercell_membership = MC$membership, </span><br><span class="line">    method = method_purity</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Metacell purity distribution</span></span><br><span class="line">summary(MC$purity)</span><br><span class="line"></span><br><span class="line">hist(MC$purity, main = paste0(<span class="string">&quot;Purity of metacells \nin terms of cell line composition (&quot;</span>, method_purity,<span class="string">&quot;)&quot;</span>))</span><br><span class="line"></span><br><span class="line">supercell_plot(</span><br><span class="line">    MC$graph.supercells, </span><br><span class="line">    group = MC$cell_line, </span><br><span class="line">    seed = <span class="number">1</span>, </span><br><span class="line">    alpha = -<span class="built_in">pi</span>/<span class="number">2</span>,</span><br><span class="line">    main  = <span class="string">&quot;Metacells colored by cell line assignment&quot;</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">supercell_plot(</span><br><span class="line">    MC$graph.singlecell, </span><br><span class="line">    group = sce$seurat_annotations, </span><br><span class="line">    do.frames = <span class="literal">FALSE</span>,</span><br><span class="line">    lay.method = <span class="string">&quot;components&quot;</span>,</span><br><span class="line">    seed = <span class="number">1</span>, </span><br><span class="line">    alpha = -<span class="built_in">pi</span>/<span class="number">2</span>,</span><br><span class="line">    main  = <span class="string">&quot;Single cells colored by cell line assignment&quot;</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment">#直接使用标准的下游处理，也即不管weight的问题</span></span><br><span class="line">MC.seurat &lt;- supercell_2_Seurat(</span><br><span class="line">    SC.GE = MC.ge, </span><br><span class="line">    SC = MC, </span><br><span class="line">    fields = <span class="built_in">c</span>(<span class="string">&quot;cell_line&quot;</span>, <span class="string">&quot;purity&quot;</span>),</span><br><span class="line">    var.genes = MC$genes.use,</span><br><span class="line">    N.comp = <span class="number">10</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">Idents(MC.seurat) &lt;- <span class="string">&quot;cell_line&quot;</span></span><br><span class="line"></span><br><span class="line">MC.seurat &lt;- RunUMAP(MC.seurat, dims = <span class="number">1</span>:<span class="number">10</span>)</span><br><span class="line"></span><br><span class="line">DimPlot(MC.seurat, reduction = <span class="string">&quot;umap&quot;</span>,pt.size = <span class="number">2.0</span>,label = <span class="built_in">T</span>) +</span><br><span class="line">    DimPlot(sce,reduction = <span class="string">&quot;umap&quot;</span>,group.by = <span class="string">&quot;seurat_annotations&quot;</span>,label = <span class="built_in">T</span>,pt.size = <span class="number">1.2</span>)</span><br><span class="line"></span><br><span class="line">VlnPlot(MC.seurat, features = <span class="built_in">c</span>(<span class="string">&quot;CD8A&quot;</span>,<span class="string">&quot;CD4&quot;</span>,<span class="string">&quot;GZMA&quot;</span>,<span class="string">&quot;GZMB&quot;</span>,<span class="string">&quot;S100A6&quot;</span>), pt.size = <span class="number">0.0</span>)</span><br></pre></td></tr></table></div></figure>
<p>从聚类结果上将，效果确实不错！而且同样兼顾了稀有细胞的问题，还支持meta信息的直接映射，支持直接使用aggregating后的数据走Seurat的流程，也提到了细胞不均等的时候引入weight的方法，考虑得非常全面，甚至支持aggregating后进行velocyto的运算，综合性能非常好，以后大规模数据就用这个了，而且这个工具aggregating后甚至支持去批次的问题，也即可以从一开始，对于每个sample的数据就进行aggregating，然后再进行下游分析，就能极大程度减少运算压力：</p>
<p>直接用TISCH2上面的CRC的<span class="exturl"><a class="exturl__link" target="_blank" rel="noopener" href="http://tisch.comp-genomics.org/gallery/?cancer=CRC&amp;species=Human">数据</a><span class="exturl__icon"><i class="fas fa-external-link-alt"></i></span></span>，来自<span class="exturl"><a class="exturl__link" target="_blank" rel="noopener" href="https://pubmed.ncbi.nlm.nih.gov/?term=32561858">文章</a><span class="exturl__icon"><i class="fas fa-external-link-alt"></i></span></span>，数据为7 patients,no treatment,10X,primary,2.3W cells，理论上这种大小的数据集当然能直接分析，但是对于这种每个样本单独获取数据的形式，如果数据扩大10倍呢？那显然就得引入一些降低开销的方法了：</p>
<figure class="highlight r"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br></pre></td><td class="code"><pre><span class="line">sce &lt;- Read10X_h5(<span class="string">&quot;./CRC_EMTAB8107_expression.h5&quot;</span>)</span><br><span class="line">sce.meta &lt;- read.table(<span class="string">&quot;./CRC_EMTAB8107_CellMetainfo_table.tsv&quot;</span>,</span><br><span class="line">                       header = <span class="built_in">T</span>,</span><br><span class="line">                       check.names = <span class="built_in">F</span>,row.names = <span class="number">1</span>,</span><br><span class="line">                       sep = <span class="string">&quot;\t&quot;</span>)</span><br><span class="line"></span><br><span class="line">sce &lt;- CreateSeuratObject(sce,</span><br><span class="line">                          project = <span class="string">&quot;CRC_EMTAB8107&quot;</span>,</span><br><span class="line">                          meta.data = sce.meta,</span><br><span class="line">                          min.features = <span class="number">200</span>,</span><br><span class="line">                          min.cells = <span class="number">10</span>)</span><br><span class="line">sce</span><br><span class="line"><span class="comment">#22236 features across 34284 samples within 1 assay</span></span><br><span class="line"></span><br><span class="line">library(gtsummary)</span><br><span class="line">table1 &lt;- tbl_summary(sce@meta.data)</span><br><span class="line">table1</span><br><span class="line"><span class="comment">#可以看到，orig.ident和sample不一致，但是sample和patient一致</span></span><br><span class="line"><span class="comment">#因此这个去批次选什么很迷惑，由于这里是癌症的数据，因此....用harmony对patient整合，然后弱去批次</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#或者其实这里就可以直接split然后做supercell</span></span><br><span class="line">sce.list &lt;- SplitObject(sce,split.by = <span class="string">&quot;Patient&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#首先单独QC</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#单独对每个seurat对象执行supercell操作，最后仍然得到一个list</span></span><br><span class="line">gamma &lt;- 20</span><br><span class="line">MC.list &lt;- <span class="built_in">list</span>()</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span>(i <span class="keyword">in</span> <span class="built_in">seq_along</span>(sce.list))&#123;</span><br><span class="line">    sce.list[[i]] &lt;- NormalizeData(sce.list[[i]])</span><br><span class="line">    sce.list[[i]] &lt;- FindVariableFeatures(sce.list[[i]])</span><br><span class="line">    hvg = VariableFeatures(sce.list[[i]])</span><br><span class="line">    MC = SCimplify(</span><br><span class="line">        X = GetAssayData(sce.list[[i]]),</span><br><span class="line">        genes.use = hvg, </span><br><span class="line">        <span class="built_in">gamma</span> = <span class="built_in">gamma</span>,</span><br><span class="line">        n.pc = <span class="number">10</span>)</span><br><span class="line">    MC.ge = supercell_GE(</span><br><span class="line">        ge = GetAssayData(sce.list[[i]]),</span><br><span class="line">        groups = MC$membership)</span><br><span class="line">    MC$Cell_line &lt;- supercell_assign(</span><br><span class="line">        cluster = sce.list[[i]]$Celltype..major.lineage.,</span><br><span class="line">        supercell_membership = MC$membership,</span><br><span class="line">        method = <span class="string">&quot;jaccard&quot;</span>)</span><br><span class="line">    MC$Sample &lt;- supercell_assign(</span><br><span class="line">        cluster = sce.list[[i]]$Sample,</span><br><span class="line">        supercell_membership = MC$membership,</span><br><span class="line">        method = <span class="string">&quot;absolute&quot;</span>)</span><br><span class="line">    MC.list[[i]] &lt;- supercell_2_Seurat(</span><br><span class="line">        SC.GE = MC.ge, </span><br><span class="line">        SC = MC, </span><br><span class="line">        fields = <span class="built_in">c</span>(<span class="string">&quot;Cell_line&quot;</span>,<span class="string">&quot;Sample&quot;</span>),</span><br><span class="line">        var.genes = MC$genes.use,</span><br><span class="line">        N.comp = <span class="number">15</span>,</span><br><span class="line">        is.log.normalized = <span class="built_in">F</span>)&#125;</span><br><span class="line">rm(MC,MC.ge)</span><br><span class="line"></span><br><span class="line">MC.harmony &lt;- merge(MC.list[[<span class="number">1</span>]], y=<span class="built_in">c</span>(MC.list[[<span class="number">2</span>]], MC.list[[<span class="number">3</span>]],MC.list[[<span class="number">4</span>]],</span><br><span class="line">                                      MC.list[[<span class="number">5</span>]], MC.list[[<span class="number">6</span>]], MC.list[[<span class="number">7</span>]]))</span><br><span class="line"></span><br><span class="line">MC.harmony &lt;- NormalizeData(MC.harmony) %&gt;% FindVariableFeatures() %&gt;% ScaleData() %&gt;% RunPCA(verbose=<span class="literal">FALSE</span>)</span><br><span class="line"></span><br><span class="line">library(harmony)</span><br><span class="line">MC.harmony &lt;- RunHarmony(MC.harmony, group.by.vars = <span class="string">&quot;Sample&quot;</span>)</span><br><span class="line"></span><br><span class="line">MC.harmony &lt;- RunUMAP(MC.harmony, reduction = <span class="string">&quot;harmony&quot;</span>, dims = <span class="number">1</span>:<span class="number">10</span>)</span><br><span class="line">MC.harmony &lt;- FindNeighbors(MC.harmony, reduction = <span class="string">&quot;harmony&quot;</span>, dims = <span class="number">1</span>:<span class="number">10</span>) %&gt;% FindClusters()</span><br><span class="line"></span><br><span class="line"><span class="comment">#group_by_cluster</span></span><br><span class="line">plot1 = DimPlot(MC.harmony, reduction = <span class="string">&quot;umap&quot;</span>,group.by = <span class="string">&quot;Cell_line&quot;</span>, label=<span class="built_in">T</span>,pt.size = <span class="number">1.2</span>) </span><br><span class="line"><span class="comment">#group_by_sample</span></span><br><span class="line">plot2 = DimPlot(MC.harmony, reduction = <span class="string">&quot;umap&quot;</span>, group.by=<span class="string">&#x27;Sample&#x27;</span>,pt.size = <span class="number">1.2</span>) </span><br><span class="line"><span class="comment">#combinate</span></span><br><span class="line">plotc &lt;- plot1+plot2</span><br><span class="line">plotc</span><br><span class="line"></span><br><span class="line"><span class="comment">#检查一下marker的情况，看看在表达上supercell是否有优势</span></span><br><span class="line">genes.to.check &lt;- <span class="built_in">c</span>(<span class="string">&#x27;MS4A1&#x27;</span>,<span class="string">&#x27;CD79A&#x27;</span>,<span class="string">&#x27;CD37&#x27;</span>,<span class="string">&#x27;CD74&#x27;</span>,</span><br><span class="line">                    <span class="string">&#x27;IL7R&#x27;</span>,<span class="string">&#x27;LTB&#x27;</span>,<span class="string">&#x27;CD3D&#x27;</span>,<span class="string">&#x27;CXCR4&#x27;</span>,</span><br><span class="line">                    <span class="string">&#x27;GNLY&#x27;</span>,<span class="string">&#x27;CCL5&#x27;</span>,<span class="string">&#x27;GZMA&#x27;</span>,<span class="string">&#x27;NKG7&#x27;</span>,</span><br><span class="line">                    <span class="string">&#x27;S100A9&#x27;</span>,<span class="string">&#x27;S100A8&#x27;</span>,<span class="string">&#x27;TYROBP&#x27;</span>,<span class="string">&#x27;LYZ&#x27;</span>,</span><br><span class="line">                    <span class="string">&#x27;JCHAIN&#x27;</span>,<span class="string">&#x27;IGKC&#x27;</span>,<span class="string">&#x27;MZB1&#x27;</span>,<span class="string">&#x27;IGLC2&#x27;</span>)</span><br><span class="line"></span><br><span class="line">SCpubr::do_DotPlot(sample = MC.harmony,</span><br><span class="line">                   features = genes.to.check,</span><br><span class="line">                   group.by = <span class="string">&quot;Cell_line&quot;</span>)</span><br><span class="line"></span><br><span class="line">SCpubr::do_CorrelationPlot(sample = MC.harmony,</span><br><span class="line">                           group.by = <span class="string">&quot;Cell_line&quot;</span>,</span><br><span class="line">                           column_names_rot = <span class="number">90</span>,</span><br><span class="line">                           cell_size = <span class="number">8</span>,</span><br><span class="line">                           cluster_cols = <span class="built_in">T</span>)</span><br><span class="line"></span><br><span class="line">network &lt;- decoupleR::get_dorothea(organism = <span class="string">&quot;human&quot;</span>,</span><br><span class="line">                                   levels = <span class="built_in">c</span>(<span class="string">&quot;A&quot;</span>, <span class="string">&quot;B&quot;</span>, <span class="string">&quot;C&quot;</span>))</span><br><span class="line"></span><br><span class="line">table(MC.harmony$Cell_line,MC.harmony$Sample)</span><br></pre></td></tr></table></div></figure>
<p>整个流程非常nice，几乎和原生单细胞流程一致，从表达量上，也能够提升marker基因的表现，harmony的结果也非常可靠，和TISCH网站上的整体数据相对比，降维聚类的结构完全一致，基因表达上则更加准确，显然是一种极好地公共数据大规模复用的方法！</p>
</div><footer class="post-footer"><div class="post-ending ending"><div class="ending__text">------ 本文结束，感谢您的阅读 ------</div></div><div class="post-copyright copyright"><div class="copyright-author"><span class="copyright-author__name">本文作者: </span><span class="copyright-author__value"><a href="https://biobrick.github.io">Chong Yin</a></span></div><div class="copyright-link"><span class="copyright-link__name">本文链接: </span><span class="copyright-link__value"><a href="https://biobrick.github.io/2023/03/02/%E5%8D%95%E7%BB%86%E8%83%9E%E4%B8%8B%E9%87%87%E6%A0%B7/">https://biobrick.github.io/2023/03/02/%E5%8D%95%E7%BB%86%E8%83%9E%E4%B8%8B%E9%87%87%E6%A0%B7/</a></span></div><div class="copyright-notice"><span class="copyright-notice__name">版权声明: </span><span class="copyright-notice__value">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en" rel="external nofollow" target="_blank">BY-NC-SA</a> 许可协议。转载请注明出处！</span></div></div><div class="post-reward reward"><div class="reward-button">请我喝杯咖啡~</div><div class="reward-qrcode"><span class="reward-qrcode-alipay"><img class="reward-qrcode-alipay__img" src="/images/alipay.png"><div class="reward-qrcode-alipay__text">支付宝打赏</div></span><span class="reward-qrcode-wechat"><img class="reward-qrcode-wechat__img" src="/images/wechatpay.png"><div class="reward-qrcode-wechat__text">微信打赏</div></span></div></div><nav class="post-paginator paginator"><div class="paginator-prev"><a class="paginator-prev__link" href="/2023/03/27/scRNA-in-R/"><span class="paginator-prev__icon"><i class="fas fa-angle-left"></i></span><span class="paginator-prev__text">scRNA in R (1)</span></a></div><div class="paginator-next"><a class="paginator-next__link" href="/2022/12/20/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E7%9A%84RNA-Velocity/"><span class="paginator-prev__text">从零开始的RNA Velocity</span><span class="paginator-next__icon"><i class="fas fa-angle-right"></i></span></a></div></nav></footer></div></div></div><div class="sidebar-wrap" id="sidebar-wrap"><aside class="sidebar" id="sidebar"><div class="sidebar-nav"><span class="sidebar-nav-toc current">文章目录</span><span class="sidebar-nav-ov">站点概览</span></div><section class="sidebar-toc"></section><!-- ov = overview--><section class="sidebar-ov hide"><div class="sidebar-ov-social"><a class="sidebar-ov-social-item" href="https://github.com/BIOBRICK/" target="_blank" rel="noopener" data-popover="Github" data-popover-pos="up"><span class="sidebar-ov-social-item__icon"><i class="fab fa-github"></i></span></a><a class="sidebar-ov-social-item" href="386396663" target="_blank" rel="noopener" data-popover="QQ" data-popover-pos="up"><span class="sidebar-ov-social-item__icon"><i class="fab fa-qq"></i></span></a></div><div class="sidebar-ov-state"><a class="sidebar-ov-state-item sidebar-ov-state-item--posts" href="/archives/"><div class="sidebar-ov-state-item__count">23</div><div class="sidebar-ov-state-item__name">归档</div></a></div><div class="sidebar-ov-cc"><a href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en" target="_blank" rel="noopener" data-popover="知识共享许可协议" data-popover-pos="up"><img src="/images/cc-by-nc-sa.svg"></a></div></section><div class="sidebar-reading"><div class="sidebar-reading-info"><span class="sidebar-reading-info__text">你已阅读了 </span><span class="sidebar-reading-info__num">0</span><span class="sidebar-reading-info__perc">%</span></div><div class="sidebar-reading-line"></div></div></aside></div><div class="clearfix"></div></div></main><footer class="footer" id="footer"><div class="footer-inner"><div><span>Copyright © 2023</span><span class="footer__icon"><i class="fas fa-heart"></i></span><span>Chong Yin</span></div><div><span>主题 - <a href="https://github.com/liuyib/hexo-theme-stun/" title="Stun" target="_blank" rel="noopener">Stun</a></span><span> v2.6.2</span></div></div></footer><div class="loading-bar" id="loading-bar"><div class="loading-bar__progress"></div></div><div class="back2top" id="back2top"><span class="back2top__icon"><i class="fas fa-rocket"></i></span></div></div><script src="https://cdn.jsdelivr.net/npm/jquery@v3.4.1/dist/jquery.min.js"></script><script src="https://cdn.jsdelivr.net/npm/velocity-animate@1.5.2/velocity.min.js"></script><script src="https://cdn.jsdelivr.net/npm/velocity-animate@1.5.2/velocity.ui.min.js"></script><script src="/js/utils.js?v=2.6.2"></script><script src="/js/stun-boot.js?v=2.6.2"></script><script src="/js/scroll.js?v=2.6.2"></script><script src="/js/header.js?v=2.6.2"></script><script src="/js/sidebar.js?v=2.6.2"></script></body></html>