<!DOCTYPE html><html lang="zh-CN"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1"><meta name="format-detection" content="telephone=no"><meta name="apple-mobile-web-app-capable" content="yes"><meta name="apple-mobile-web-app-status-bar-style" content="black"><link rel="icon" href="/images/icons/favicon-16x16.png?v=2.6.2" type="image/png" sizes="16x16"><link rel="icon" href="/images/icons/favicon-32x32.png?v=2.6.2" type="image/png" sizes="32x32"><meta name="description" content="单细胞的上游流程，使用了kallisto替代cellranger，下游使用scanpy，尝试在python中搭建完整的单细胞分析栈。">
<meta property="og:type" content="article">
<meta property="og:title" content="单细胞kallisto传统流程">
<meta property="og:url" content="https://biobrick.github.io/2021/12/04/%E5%8D%95%E7%BB%86%E8%83%9Ekallisto%E4%BC%A0%E7%BB%9F%E6%B5%81%E7%A8%8B/">
<meta property="og:site_name" content="Bio Harbor">
<meta property="og:description" content="单细胞的上游流程，使用了kallisto替代cellranger，下游使用scanpy，尝试在python中搭建完整的单细胞分析栈。">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2021-12-03T18:59:54.000Z">
<meta property="article:modified_time" content="2021-12-03T19:27:11.094Z">
<meta property="article:author" content="Chong Yin">
<meta name="twitter:card" content="summary"><title>单细胞kallisto传统流程 | Bio Harbor</title><link ref="canonical" href="https://biobrick.github.io/2021/12/04/%E5%8D%95%E7%BB%86%E8%83%9Ekallisto%E4%BC%A0%E7%BB%9F%E6%B5%81%E7%A8%8B/"><link rel="dns-prefetch" href="https://cdn.jsdelivr.net"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.12.1/css/all.min.css" type="text/css"><link rel="stylesheet" href="/css/index.css?v=2.6.2"><script>var Stun = window.Stun || {};
var CONFIG = {
  root: '/',
  algolia: undefined,
  assistSearch: undefined,
  fontIcon: {"prompt":{"success":"fas fa-check-circle","info":"fas fa-arrow-circle-right","warning":"fas fa-exclamation-circle","error":"fas fa-times-circle"},"copyBtn":"fas fa-copy"},
  sidebar: {"offsetTop":"20px","tocMaxDepth":6},
  header: {"enable":true,"showOnPost":true,"scrollDownIcon":false},
  postWidget: {"endText":true},
  nightMode: {"enable":true},
  back2top: {"enable":true},
  codeblock: {"style":"default","highlight":"light","wordWrap":false},
  reward: false,
  fancybox: false,
  zoomImage: {"gapAside":"20px"},
  galleryWaterfall: undefined,
  lazyload: false,
  pjax: undefined,
  externalLink: {"icon":{"enable":true,"name":"fas fa-external-link-alt"}},
  shortcuts: undefined,
  prompt: {"copyButton":"复制","copySuccess":"复制成功","copyError":"复制失败"},
  sourcePath: {"js":"js","css":"css","images":"images"},
};

window.CONFIG = CONFIG;</script><meta name="generator" content="Hexo 5.4.0"></head><body><div class="container" id="container"><header class="header" id="header"><div class="header-inner"><nav class="header-nav header-nav--fixed"><div class="header-nav-inner"><div class="header-nav-menubtn"><i class="fas fa-bars"></i></div><div class="header-nav-menu"><div class="header-nav-menu-item"><a class="header-nav-menu-item__link" href="/"><span class="header-nav-menu-item__icon"><i class="fas fa-home"></i></span><span class="header-nav-menu-item__text">首页</span></a></div><div class="header-nav-menu-item"><a class="header-nav-menu-item__link" href="/archives/"><span class="header-nav-menu-item__icon"><i class="fas fa-folder-open"></i></span><span class="header-nav-menu-item__text">归档</span></a></div></div><div class="header-nav-mode"><div class="mode"><div class="mode-track"><span class="mode-track-moon"></span><span class="mode-track-sun"></span></div><div class="mode-thumb"></div></div></div></div></nav><div class="header-banner"><div class="header-banner-info"><div class="header-banner-info__title">Bio Harbor</div><div class="header-banner-info__subtitle"></div></div></div></div></header><main class="main" id="main"><div class="main-inner"><div class="content-wrap" id="content-wrap"><div class="content" id="content"><!-- Just used to judge whether it is an article page--><div id="is-post"></div><div class="post"><header class="post-header"><h1 class="post-title">单细胞kallisto传统流程</h1><div class="post-meta"><span class="post-meta-item post-meta-item--createtime"><span class="post-meta-item__icon"><i class="far fa-calendar-plus"></i></span><span class="post-meta-item__info">发表于</span><span class="post-meta-item__value">2021-12-04</span></span><span class="post-meta-item post-meta-item--updatetime"><span class="post-meta-item__icon"><i class="far fa-calendar-check"></i></span><span class="post-meta-item__info">更新于</span><span class="post-meta-item__value">2021-12-04</span></span></div></header><div class="post-body">
        <h2 id="kallisto-bustool-scanpy流程"   >
          <a href="#kallisto-bustool-scanpy流程" class="heading-link"><i class="fas fa-link"></i></a><a href="#kallisto-bustool-scanpy流程" class="headerlink" title="kallisto/bustool/scanpy流程"></a>kallisto/bustool/scanpy流程</h2>
      <p>通常，10X的上游都是cellranger，但是这玩意消耗大，速度慢，而且定制化程度很高，cellranger虽然是集成的STAR的流程，但是下游结果文件固定，参数也很多。个人用起来太重量级了。因此，有了某些大佬，决定针对scRNA的实际需求，改进了kallisto工具，并结合高端的bustool，因此就产生了这个上游kallisto+bustools，下游scanpy+其他高级可视化的workflow。</p>

        <h2 id="Node1-data"   >
          <a href="#Node1-data" class="heading-link"><i class="fas fa-link"></i></a><a href="#Node1-data" class="headerlink" title="Node1.data"></a>Node1.data</h2>
      <p>10X官网下载scRNA的测序数据即可，只需要输入一些基本信息，即可随意下载，根据自己的电脑配置下载，我这里下载的是1K PBMC的，大了放不下，最近动漫下多了。下载数据没什么特别的，但是需要注意，文库版本是哪个版本的，10X的V2文库和V3文库差别很大，而且这个参数后面需要。</p>

        <h2 id="Node2-Ref"   >
          <a href="#Node2-Ref" class="heading-link"><i class="fas fa-link"></i></a><a href="#Node2-Ref" class="headerlink" title="Node2.Ref"></a>Node2.Ref</h2>
      <p>kallisto是把fastq与<strong>转录组</strong>参考相对比，因此需要的并不是传统的转录组下载基因组的参考，而是需要去ensembl下载转录组的文件，注意，转录组，转录组，转录组，几个文件注意版本对应分别为：</p>
<figure class="highlight shell"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Homo_sapiens.GRCh38.cdna.all.fa.gz #这个是cDNA的，也就是转录本参考</span><br><span class="line">Homo_sapiens.GRCh38.104.gtf.gz #注意要去转录本的文件夹里下载转录本的gtf</span><br><span class="line">Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz #这个是基因组的</span><br></pre></td></tr></table></div></figure>

<p>借助一个叫kb-python的warpper，可以很容易的使用kallisto和bustool的联合，这里贴出官方的参数：</p>
<figure class="highlight bash"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br></pre></td><td class="code"><pre><span class="line">$ kb ref</span><br><span class="line">usage: kb ref [-h] [--tmp TMP] [--keep-tmp] [--verbose] -i INDEX -g T2G -f1</span><br><span class="line">              FASTA [-f2 FASTA] [-c1 T2C] [-c2 T2C] [-n N]</span><br><span class="line">              [-d &#123;human,mouse,linnarsson&#125;] [-k K]</span><br><span class="line">              [--workflow &#123;standard,lamanno,nucleus,kite&#125;] [--lamanno]</span><br><span class="line">              [--overwrite]</span><br><span class="line">              fasta gtf [feature]</span><br><span class="line"></span><br><span class="line">Build a kallisto index and transcript-to-gene mapping</span><br><span class="line"></span><br><span class="line">positional arguments:</span><br><span class="line">  fasta                 Genomic FASTA file(s), comma-delimited</span><br><span class="line">  gtf                   Reference GTF file(s), comma-delimited</span><br><span class="line">  feature               [`kite` workflow only] Path to TSV containing barcodes</span><br><span class="line">                        and feature names.</span><br><span class="line"></span><br><span class="line">optional arguments:</span><br><span class="line">  -h, --<span class="built_in">help</span>            Show this <span class="built_in">help</span> message and <span class="built_in">exit</span></span><br><span class="line">  --tmp TMP             Override default temporary directory</span><br><span class="line">  --keep-tmp            Do not delete the tmp directory</span><br><span class="line">  --verbose             Print debugging information</span><br><span class="line">  -n N                  Number of files to split the index into. If this</span><br><span class="line">                        option is specified, the FASTA that is normally used</span><br><span class="line">                        to create an index is split into `N` approximately-</span><br><span class="line">                        equal parts. Each of these FASTAs are indexed</span><br><span class="line">                        separately.</span><br><span class="line">  -d &#123;human,mouse,linnarsson&#125;</span><br><span class="line">                        Download a pre-built kallisto index (along with all</span><br><span class="line">                        necessary files) instead of building it locally</span><br><span class="line">  -k K                  Use this option to override the k-mer length of the</span><br><span class="line">                        index. Usually, the k-mer length automatically</span><br><span class="line">                        calculated by `kb` provides the best results.</span><br><span class="line">  --workflow &#123;standard,lamanno,nucleus,kite&#125;</span><br><span class="line">                        Type of workflow to prepare files <span class="keyword">for</span>. Use `lamanno`</span><br><span class="line">                        <span class="keyword">for</span> RNA velocity based on La Manno et al. 2018 logic.</span><br><span class="line">                        Use `nucleus` <span class="keyword">for</span> RNA velocity on single-nucleus RNA-</span><br><span class="line">                        seq reads. Use `kite` <span class="keyword">for</span> feature barcoding. (default:</span><br><span class="line">                        standard)</span><br><span class="line">  --lamanno             Deprecated. Use `--workflow lamanno` instead.</span><br><span class="line">  --overwrite           Overwrite existing kallisto index</span><br><span class="line"></span><br><span class="line">required arguments:</span><br><span class="line">  -i INDEX              Path to the kallisto index to be constructed. If `-n`</span><br><span class="line">                        is also specified, this is the prefix <span class="keyword">for</span> the n</span><br><span class="line">                        indices to construct.</span><br><span class="line">  -g T2G                Path to transcript-to-gene mapping to be generated</span><br><span class="line">  -f1 FASTA             [Optional with -d] Path to the cDNA FASTA (lamanno,</span><br><span class="line">                        nucleus) or mismatch FASTA (kite) to be generated</span><br><span class="line"></span><br><span class="line">required arguments <span class="keyword">for</span> `lamanno` and `nucleus` workflows:</span><br><span class="line">  -f2 FASTA             Path to the intron FASTA to be generated</span><br><span class="line">  -c1 T2C               Path to generate cDNA transcripts-to-capture</span><br><span class="line">  -c2 T2C               Path to generate intron transcripts-to-capture</span><br></pre></td></tr></table></div></figure>

<p>特别注意，如果在一开始加了-d参数，那么会从一个在线存储里面下载预先已经建立好的index，然而不出所料是下载不了的。根据后面的教程，如果要走standard的单细胞流程，需要2个文件，一个idx文件，就是索引，和一个t2g文件，transcripts-to-capture，就是一个基因注释文件，大概长这样，理论上这东西直接自己提取也不是不行……</p>
<figure class="highlight plaintext"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">NST00000456328.2    ENSG00000223972.5    DDX11L1</span><br><span class="line">ENST00000450305.2    ENSG00000223972.5    DDX11L1</span><br><span class="line">ENST00000488147.1    ENSG00000227232.5    WASH7P</span><br><span class="line">ENST00000619216.1    ENSG00000278267.1    MIR6859-1</span><br><span class="line">ENST00000473358.1    ENSG00000243485.5    MIR1302-2HG</span><br><span class="line">ENST00000469289.1    ENSG00000243485.5    MIR1302-2HG</span><br><span class="line">ENST00000607096.1    ENSG00000284332.1    MIR1302-2</span><br><span class="line">ENST00000417324.1    ENSG00000237613.2    FAM138A</span><br><span class="line">ENST00000461467.1    ENSG00000237613.2    FAM138A</span><br><span class="line">ENST00000606857.1    ENSG00000268020.3    OR4G4P</span><br><span class="line">ENST00000642116.1    ENSG00000240361.2    OR4G11P</span><br><span class="line">ENST00000492842.2    ENSG00000240361.2    OR4G11P</span><br><span class="line">ENST00000641515.2    ENSG00000186092.6    OR4F5</span><br><span class="line">ENST00000335137.4    ENSG00000186092.6    OR4F5</span><br></pre></td></tr></table></div></figure>

<p>还要注意，里面有个workflow参数，是用来指定对应流程的，这个教程先不考虑其他，直接走最经典的单细胞转录组流程，那么这个参数可以不动，默认就是standard，但是，如果准备走最早的velocity流程，那么需要使用lamanno，如果需要走单细胞核测序的RNA velocity（就是scVelo），需要使用nucleus，而要走10X的Feature Barcoding（测表面蛋白）的话，则需要使用kite。</p>
<p>这里我们走经典转录组，而且不需要太复杂，因此脚本可以为stdindex.sh</p>
<figure class="highlight bash"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#!/usr/bin</span></span><br><span class="line"></span><br><span class="line">transindexpath=../refdata/standard/trans104.idx</span><br><span class="line">g2fpath=../refdata/standard/g2f.txt</span><br><span class="line">transfastapath=<span class="string">&quot;../refdata/standard/cDNA.fa</span></span><br><span class="line"><span class="string">genomepath=../refdata/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz</span></span><br><span class="line"><span class="string">gtfpath=../refdata/Homo_sapiens.GRCh38.104.gtf.gz</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">kb ref -i <span class="variable">$transindexpath</span> -g <span class="variable">$g2fpath</span> -f1 <span class="variable">$transfastapath</span> <span class="variable">$fastapath</span> <span class="variable">$gtfpath</span></span></span><br></pre></td></tr></table></div></figure>

<p>运行大概持续了15min，速度还是相当快的，而且结果和直接下载得到的cdna.all.fa和idex文件基本一致，当然因为版本原因大小有一点区别的还是。</p>
<p>这里需要注意，如果我们下载了cdna.all.fa文件，最后输出的时候如果同文件夹下存在的话，会跳过这一步的输出。最后idx大小大概是3G多，g2f文件10多M，cDNA文件400多M。</p>
<p>所以，这个就是经典scRNA-seq流程的准备文件了，下面就是准备进入分析流程了。</p>

        <h2 id="Node3-count"   >
          <a href="#Node3-count" class="heading-link"><i class="fas fa-link"></i></a><a href="#Node3-count" class="headerlink" title="Node3.count"></a>Node3.count</h2>
      <p>这一步仍然是利用kb-python来进行比对和计数，由于kb-python的封装，这一步之后就能直接出所有的下游需要的文件了。以及…其实kb-python本身也就只有2个外部接口罢了。</p>
<p>这里也给出kb count的参数，可以看到，这里多了很多就是了</p>
<figure class="highlight bash"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br></pre></td><td class="code"><pre><span class="line">usage: kb count [-h] [--tmp TMP] [--keep-tmp] [--verbose] -i INDEX -g T2G -x</span><br><span class="line">                TECHNOLOGY [-o OUT] [-w WHITELIST] [-t THREADS] [-m MEMORY]</span><br><span class="line">                [--workflow &#123;standard,lamanno,nucleus,kite,kite:10xFB&#125;]</span><br><span class="line">                [--mm | --tcc] [--filter [&#123;bustools&#125;]] [-c1 T2C] [-c2 T2C]</span><br><span class="line">                [--overwrite] [--dry-run] [--lamanno | --nucleus]</span><br><span class="line">                [--loom | --h5ad]</span><br><span class="line">                fastqs [fastqs ...]</span><br><span class="line"></span><br><span class="line">Generate count matrices from a <span class="built_in">set</span> of single-cell FASTQ files. Run `kb --list`</span><br><span class="line">to view single-cell technology information.</span><br><span class="line"></span><br><span class="line">positional arguments:</span><br><span class="line">  fastqs                FASTQ files</span><br><span class="line"></span><br><span class="line">optional arguments:</span><br><span class="line">  -h, --<span class="built_in">help</span>            Show this <span class="built_in">help</span> message and <span class="built_in">exit</span></span><br><span class="line">  --tmp TMP             Override default temporary directory</span><br><span class="line">  --keep-tmp            Do not delete the tmp directory</span><br><span class="line">  --verbose             Print debugging information</span><br><span class="line">  -o OUT                Path to output directory (default: current directory)</span><br><span class="line">  -w WHITELIST          Path to file of whitelisted barcodes to correct to. If</span><br><span class="line">                        not provided and bustools supports the technology, a</span><br><span class="line">                        pre-packaged whitelist is used. If not, the bustools</span><br><span class="line">                        whitelist <span class="built_in">command</span> is used. (`kb --list` to view</span><br><span class="line">                        whitelists)</span><br><span class="line">  -t THREADS            Number of threads to use (default: 8)</span><br><span class="line">  -m MEMORY             Maximum memory used (default: 4G)</span><br><span class="line">  --workflow &#123;standard,lamanno,nucleus,kite,kite:10xFB&#125;</span><br><span class="line">                        Type of workflow. Use `lamanno` <span class="keyword">for</span> RNA velocity based</span><br><span class="line">                        on La Manno et al. 2018 logic. Use `nucleus` <span class="keyword">for</span> RNA</span><br><span class="line">                        velocity on single-nucleus RNA-seq reads. Use `kite`</span><br><span class="line">                        <span class="keyword">for</span> feature barcoding. Use `kite:10xFB` <span class="keyword">for</span> 10x</span><br><span class="line">                        Genomics Feature Barcoding technology. (default:</span><br><span class="line">                        standard)</span><br><span class="line">  --mm                  Include reads that pseudoalign to multiple genes.</span><br><span class="line">  --tcc                 Generate a TCC matrix instead of a gene count matrix.</span><br><span class="line">  --filter [&#123;bustools&#125;]</span><br><span class="line">                        Produce a filtered gene count matrix (default:</span><br><span class="line">                        bustools)</span><br><span class="line">  --overwrite           Overwrite existing output.bus file</span><br><span class="line">  --dry-run             Dry run</span><br><span class="line">  --lamanno             Deprecated. Use `--workflow lamanno` instead.</span><br><span class="line">  --nucleus             Deprecated. Use `--workflow nucleus` instead.</span><br><span class="line">  --loom                Generate loom file from count matrix</span><br><span class="line">  --h5ad                Generate h5ad file from count matrix</span><br><span class="line"></span><br><span class="line">required arguments:</span><br><span class="line">  -i INDEX              Path to kallisto index/indices, comma-delimited</span><br><span class="line">  -g T2G                Path to transcript-to-gene mapping</span><br><span class="line">  -x TECHNOLOGY         Single-cell technology used (`kb --list` to view)</span><br><span class="line"></span><br><span class="line">required arguments <span class="keyword">for</span> `lamanno` and `nucleus` workflows:</span><br><span class="line">  -c1 T2C               Path to cDNA transcripts-to-capture</span><br><span class="line">  -c2 T2C               Path to intron transcripts-to-captured</span><br></pre></td></tr></table></div></figure>

<p>同样，可以使用stdcount.sh来进行这一步:</p>
<figure class="highlight bash"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#!/usr/bin</span></span><br><span class="line"></span><br><span class="line">fastqpath=../refdata/standard/trans104.idx</span><br><span class="line">g2fpath=../refdata/standard/g2f.txt</span><br><span class="line"></span><br><span class="line">kb count --h5ad -i <span class="variable">$fastqpath</span> -g <span class="variable">$g2fpath</span> -x 10xv3 -o output --filter bustools -t 2 data/*_001.fastq.gz</span><br></pre></td></tr></table></div></figure>

<p>这里其实挺奇怪的，如果用过cellranger就能知道，用cellranger进行每个lane比对的时候是需要I1，R1和R2三个文件才能进行比对的，但是其实kallisto或者单独的STAR solo以及其他一些工具在做这一步的时候反而不需要I1那个文件，只需要R1和R2就行。</p>
<p>先看看这三个文件是什么：</p>
<p>“The source material consists of 6 FASTQ files split into two sequencing lanes <em>L001</em> and <em>L002</em>, each with three reads of <strong>R1</strong> (barcodes), <strong>R2</strong> (cDNA sequences), <strong>I1</strong> (illumina lane info)”</p>
<p>这三个文件的来源是下机的原始文件经过mkfastq或者illumina的bcl2fastq出来的，那么，需要R1和R2就很容易理解，barcode和cDNA自然是最需要的。但是为什么可以不要lane info？而且，上面的stdcount.sh会一并将lane info文件进入比对流程，但是仍然能够正常比对，尝试去掉lane info文件后再比对，也能比对出来。</p>
<p>但是，如果加入I1文件的话，比对的结果只有600M左右，但不加上I1文件则与官网给出的结果差不多，有1.3G，首先正确性上来说，肯定是不加I1文件是正确的结果，可以先看看这个所谓的lane info是什么：</p>
<figure class="highlight bash"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">head pbmc_1k_v3_S1_L001_I1_001.fastq</span><br><span class="line"></span><br><span class="line">@A00228:279:HFWFVDMXX:1:1101:8486:1000 1:N:0:NCATTACT</span><br><span class="line">NCATTACT</span><br><span class="line">+</span><br><span class="line"><span class="comment">#FFFFFFF</span></span><br><span class="line">@A00228:279:HFWFVDMXX:1:1101:10782:1000 1:N:0:NCATTACT</span><br><span class="line">NCATTACT</span><br><span class="line">+</span><br><span class="line"><span class="comment">#FFFFFFF</span></span><br><span class="line">@A00228:279:HFWFVDMXX:1:1101:12626:1000 1:N:0:NCATTACT</span><br><span class="line">NCATTACT</span><br><span class="line">+</span><br><span class="line"><span class="comment">#FFFFFFF</span></span><br><span class="line"></span><br><span class="line">tail pbmc_1k_v3_S1_L001_I1_001.fastq</span><br><span class="line">@A00228:279:HFWFVDMXX:1:2488:9670:37059 1:N:0:GGCAATGG</span><br><span class="line">GGCAATGG</span><br><span class="line">+</span><br><span class="line">FFFFFFFF</span><br><span class="line">@A00228:279:HFWFVDMXX:1:2488:13467:37059 1:N:0:GGCAATGG</span><br><span class="line">GGCAATGG</span><br><span class="line">+</span><br><span class="line">FFFFFFFF</span><br></pre></td></tr></table></div></figure>

<p>首先能看到，基本上全部是一样的东西，而且一个特征是，这个里面是不带有任何序列信息的，唯一长得像的就是比如NCATTACT，GGCAATGG这种了，那么这个8碱基的东西是啥呢。</p>
<p>根据10X V3的建库的官方说明以及提供的Single_Index_Kit_T_Set_A.csv这个文件，这个8碱基的东西应该是所谓的sample_index，一般就是使用的上面那个文件里面提供的官方的sample index序列，当然，自己指定也行，这个东西的作用很简答，就是用来区分样本用的。</p>
<p>根据下游的分析，大致能够判断，将sample index一起放进去比对，对上游比对结果影响不是特别大，当然，是有影响的，比如使用bustools inspect查看比对结果时能发现，将index一起放进比对的结果里面mean reads per barcode少了不少，甚至已经低于V3的标准（2W reads per barcode），理论上结果应当以不加入sample index的结果为准。</p>
<p>而且，从最后聚类的结果也能明显发现，不加sample index的时候，聚类结果和scanpy的官网示例基本一致，都是聚类为0-7，但是加了sample index则变成了0-9，可以认为加了sample index进比对流程的话，会影响比对出来的gene的count数。</p>
<p>总之，最后count的结果，就是output文件夹里的东西，大概有这些：</p>
<figure class="highlight bash"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">output</span><br><span class="line">├── 10x_version3_whitelist.txt</span><br><span class="line">├── counts_filtered</span><br><span class="line">│   ├── adata.h5ad</span><br><span class="line">│   ├── cells_x_genes.barcodes.txt</span><br><span class="line">│   ├── cells_x_genes.genes.txt</span><br><span class="line">│   └── cells_x_genes.mtx</span><br><span class="line">├── counts_unfiltered</span><br><span class="line">│   ├── adata.h5ad</span><br><span class="line">│   ├── cells_x_genes.barcodes.txt</span><br><span class="line">│   ├── cells_x_genes.genes.txt</span><br><span class="line">│   └── cells_x_genes.mtx</span><br><span class="line">├── filter_barcodes.txt</span><br><span class="line">├── inspect.json</span><br><span class="line">├── kb_info.json</span><br><span class="line">├── matrix.ec</span><br><span class="line">├── output.bus</span><br><span class="line">├── output.filtered.bus</span><br><span class="line">├── output.unfiltered.bus</span><br><span class="line">├── run_info.json</span><br><span class="line">└── transcripts.txt</span><br><span class="line"></span><br><span class="line">2 directories, 18 files</span><br></pre></td></tr></table></div></figure>

<p>其中bus文件是比对后通过bustools压缩的文件，而标准的10X流程则只需要counts_unfiltered/filtered中的h5ad文件或者经典三文件即可。</p>

        <h2 id="Node4-DownStream-Data-Preprocess"   >
          <a href="#Node4-DownStream-Data-Preprocess" class="heading-link"><i class="fas fa-link"></i></a><a href="#Node4-DownStream-Data-Preprocess" class="headerlink" title="Node4.DownStream-Data Preprocess"></a>Node4.DownStream-Data Preprocess</h2>
      <p>那么，继续往下，以不加sample index的结果为准，下游就能进入scanpy或者scater的标准QC过程了，这里以scanpy为主：</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> anndata</span><br><span class="line"><span class="keyword">from</span> sklearn.decomposition <span class="keyword">import</span> TruncatedSVD</span><br><span class="line"><span class="keyword">import</span> matplotlib</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"></span><br><span class="line">sc.settings.verbosity = <span class="number">3</span></span><br><span class="line">sc.settings.set_figure_params(dpi=<span class="number">80</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">adata = anndata.read_h5ad(<span class="string">&quot;output/counts_unfiltered/adata.h5ad&quot;</span>)</span><br><span class="line">adata.var[<span class="string">&quot;gene_id&quot;</span>] = adata.var.index.values</span><br><span class="line"></span><br><span class="line">t2g = pd.read_table(<span class="string">&quot;../refdata/standard/g2f.txt&quot;</span>, header=<span class="literal">None</span>, names=[<span class="string">&quot;tid&quot;</span>, <span class="string">&quot;gene_id&quot;</span>, <span class="string">&quot;gene_name&quot;</span>,<span class="string">&quot;version_name&quot;</span>,<span class="string">&quot;chr&quot;</span>,<span class="string">&quot;start&quot;</span>,<span class="string">&quot;end&quot;</span>,<span class="string">&quot;strand&quot;</span>], sep=<span class="string">&quot;\t&quot;</span>)</span><br><span class="line">t2g = t2g.dropna(axis=<span class="number">0</span>)</span><br><span class="line">t2g.index = t2g.gene_id</span><br><span class="line">t2g = t2g.loc[~t2g.index.duplicated(keep=<span class="string">&#x27;first&#x27;</span>)]</span><br><span class="line"></span><br><span class="line">adata.var[<span class="string">&quot;gene_name&quot;</span>] = adata.var.gene_id.<span class="built_in">map</span>(t2g[<span class="string">&quot;gene_name&quot;</span>])</span><br><span class="line">adata.var = adata.var.fillna(<span class="string">&quot;unknowngene&quot;</span>)</span><br><span class="line">adata.var.index = adata.var[<span class="string">&quot;gene_name&quot;</span>].tolist()</span><br><span class="line"></span><br><span class="line">adata.var_names_make_unique()</span><br></pre></td></tr></table></div></figure>

<p>这里有2个超级坑，其一，是如果使用最新版本的基因组和gtf文件，可能出现有个ensembl-id没有对应gene_name的情况，这很正常，但是进入anndata会导致后续无法分析，而且scannpy又没有提供去除Na或者空值之类的接口，因此，简单起见，如果使用了版本较高的基因组和gtf文件，务必检查前面ref的时候的那个g2f.txt里面又没有缺失或者空值，如果有，务必对anndata里面的var进行处理，我这里用的是一个比较粗暴的办法，直接把缺失设置为unknowngene，后面能够正常进行。</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adata.var = adata.var.fillna(<span class="string">&quot;unknowngene&quot;</span>)</span><br></pre></td></tr></table></div></figure>

<p>其二，在设置var的index的时候，必须如下设置：</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adata.var.index = adata.var[<span class="string">&quot;gene_name&quot;</span>].tolist()</span><br></pre></td></tr></table></div></figure>

<p>官网教程上要么是给的干净数据，要么是之前版本的接口，没有对index对象做tolist处理，这样会导致anndata的var.index设置错误，进一步导致后续部分步骤和所有的结果无法保存，而且一旦session关闭，这次处理过后的h5ad文件将无法再被读入python，等同于要重新跑。</p>
<p>继续，读入h5ad之后，就构成了一个叫anndata的对象，这个对象跟R里面的# SummarizedExperiment/Singlecellexperiment对象特别像。根据官方文档，anndata至少有3个部分</p>
<p>anndata.var：主要是基因级别信息，比如上面的gene_name这些。</p>
<p>anndata.obs：主要是cell级别的信息，比如barcode，group这些</p>
<p>anndata.X：核心数据，用于存储稀疏矩阵</p>
<p>比如在最后分析结束后，adata里面有这些东西：</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">adata</span><br><span class="line"></span><br><span class="line">AnnData <span class="built_in">object</span> <span class="keyword">with</span> n_obs × n_vars = <span class="number">1175</span> × <span class="number">25834</span></span><br><span class="line">    obs: <span class="string">&#x27;n_genes&#x27;</span>, <span class="string">&#x27;percent_mito&#x27;</span>, <span class="string">&#x27;n_counts&#x27;</span>, <span class="string">&#x27;louvain&#x27;</span>, <span class="string">&#x27;leiden&#x27;</span></span><br><span class="line">    var: <span class="string">&#x27;gene_name&#x27;</span>, <span class="string">&#x27;gene_id&#x27;</span>, <span class="string">&#x27;n_cells&#x27;</span>, <span class="string">&#x27;highly_variable&#x27;</span>, <span class="string">&#x27;means&#x27;</span>, <span class="string">&#x27;dispersions&#x27;</span>, <span class="string">&#x27;dispersions_norm&#x27;</span>, <span class="string">&#x27;mean&#x27;</span>, <span class="string">&#x27;std&#x27;</span></span><br><span class="line">    uns: <span class="string">&#x27;log1p&#x27;</span>, <span class="string">&#x27;hvg&#x27;</span>, <span class="string">&#x27;pca&#x27;</span>, <span class="string">&#x27;neighbors&#x27;</span>, <span class="string">&#x27;louvain&#x27;</span>, <span class="string">&#x27;paga&#x27;</span>, <span class="string">&#x27;louvain_sizes&#x27;</span>, <span class="string">&#x27;umap&#x27;</span>, <span class="string">&#x27;leiden&#x27;</span>, <span class="string">&#x27;leiden_colors&#x27;</span>, <span class="string">&#x27;louvain_colors&#x27;</span></span><br><span class="line">    obsm: <span class="string">&#x27;X_pca&#x27;</span>, <span class="string">&#x27;X_umap&#x27;</span></span><br><span class="line">    varm: <span class="string">&#x27;PCs&#x27;</span></span><br><span class="line">    obsp: <span class="string">&#x27;distances&#x27;</span>, <span class="string">&#x27;connectivities&#x27;</span></span><br></pre></td></tr></table></div></figure>

<p>可以看到，还有一些没说的属性，比如uns，这里存放了一些不能被结构话的数据，比如每个group绘图的时候用什么颜色映射，paga的路径之类的，里面大概长这样:</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">adata.uns</span><br><span class="line"></span><br><span class="line">OverloadedDict, wrapping:</span><br><span class="line">    OrderedDict([(<span class="string">&#x27;log1p&#x27;</span>, &#123;<span class="string">&#x27;base&#x27;</span>: <span class="literal">None</span>&#125;), (<span class="string">&#x27;hvg&#x27;</span>, &#123;<span class="string">&#x27;flavor&#x27;</span>: <span class="string">&#x27;seurat&#x27;</span>&#125;), (<span class="string">&#x27;pca&#x27;</span>, &#123;<span class="string">&#x27;params&#x27;</span>: &#123;<span class="string">&#x27;zero_center&#x27;</span>: <span class="literal">True</span>, <span class="string">&#x27;use_highly_variable&#x27;</span>: <span class="literal">True</span>&#125;, <span class="string">&#x27;variance&#x27;</span>: array([<span class="number">66.33609</span>  , <span class="number">48.00583</span>  , <span class="number">43.741863</span> , <span class="number">24.337816</span> , <span class="number">11.7938385</span>,</span><br><span class="line">        <span class="number">9.654164</span> ,  <span class="number">8.945425</span> ,  <span class="number">8.060145</span> ,  <span class="number">7.9558783</span>,  <span class="number">7.075499</span> ,</span><br><span class="line">        <span class="number">6.960134</span> ,  <span class="number">6.713339</span> ,  <span class="number">6.5481</span>   ,  <span class="number">6.3729434</span>,  <span class="number">6.224372</span> ,</span><br><span class="line">        <span class="number">6.066858</span> ,  <span class="number">5.907557</span> ,  <span class="number">5.595799</span> ,  <span class="number">5.498731</span> ,  <span class="number">5.339343</span> ,</span><br><span class="line">        <span class="number">5.264697</span> ,  <span class="number">5.085398</span> ,  <span class="number">5.0702558</span>,  <span class="number">4.9665728</span>,  <span class="number">4.9449844</span>,</span><br><span class="line">        <span class="number">4.8837957</span>,  <span class="number">4.7512255</span>,  <span class="number">4.7258067</span>,  <span class="number">4.6373615</span>,  <span class="number">4.589517</span> ,</span><br><span class="line">        <span class="number">4.484859</span> ,  <span class="number">4.4121437</span>,  <span class="number">4.3468637</span>,  <span class="number">4.294298</span> ,  <span class="number">4.2609587</span>,</span><br><span class="line">        <span class="number">4.2075477</span>,  <span class="number">4.201293</span> ,  <span class="number">4.138263</span> ,  <span class="number">4.093854</span> ,  <span class="number">4.079861</span> ,</span><br><span class="line">        <span class="number">4.035178</span> ,  <span class="number">4.023861</span> ,  <span class="number">3.9789526</span>,  <span class="number">3.9368029</span>,  <span class="number">3.9152558</span>,</span><br><span class="line">        <span class="number">3.8964837</span>,  <span class="number">3.8508425</span>,  <span class="number">3.766862</span> ,  <span class="number">3.7391431</span>,  <span class="number">3.7227194</span>],</span><br><span class="line">      dtype=float32), <span class="string">&#x27;variance_ratio&#x27;</span>: array([<span class="number">0.04367469</span>, <span class="number">0.03160632</span>, <span class="number">0.02879899</span>, <span class="number">0.01602365</span>, <span class="number">0.00776489</span>,</span><br><span class="line">       <span class="number">0.00635616</span>, <span class="number">0.00588953</span>, <span class="number">0.00530668</span>, <span class="number">0.00523803</span>, <span class="number">0.0046584</span> ,</span><br><span class="line">       <span class="number">0.00458245</span>, <span class="number">0.00441996</span>, <span class="number">0.00431117</span>, <span class="number">0.00419585</span>, <span class="number">0.00409803</span>,</span><br><span class="line">       <span class="number">0.00399433</span>, <span class="number">0.00388945</span>, <span class="number">0.00368419</span>, <span class="number">0.00362028</span>, <span class="number">0.00351534</span>,</span><br><span class="line">       <span class="number">0.0034662</span> , <span class="number">0.00334815</span>, <span class="number">0.00333818</span>, <span class="number">0.00326992</span>, <span class="number">0.0032557</span> ,</span><br><span class="line">       <span class="number">0.00321542</span>, <span class="number">0.00312814</span>, <span class="number">0.0031114</span> , <span class="number">0.00305317</span>, <span class="number">0.00302167</span>,</span><br><span class="line">       <span class="number">0.00295276</span>, <span class="number">0.00290489</span>, <span class="number">0.00286191</span>, <span class="number">0.0028273</span> , <span class="number">0.00280535</span>,</span><br><span class="line">       <span class="number">0.00277019</span>, <span class="number">0.00276607</span>, <span class="number">0.00272457</span>, <span class="number">0.00269533</span>, <span class="number">0.00268612</span>,</span><br><span class="line">       <span class="number">0.0026567</span> , <span class="number">0.00264925</span>, <span class="number">0.00261968</span>, <span class="number">0.00259193</span>, <span class="number">0.00257775</span>,</span><br><span class="line">       <span class="number">0.00256539</span>, <span class="number">0.00253534</span>, <span class="number">0.00248005</span>, <span class="number">0.0024618</span> , <span class="number">0.00245098</span>],</span><br><span class="line">      dtype=float32)&#125;), (<span class="string">&#x27;neighbors&#x27;</span>, &#123;<span class="string">&#x27;connectivities_key&#x27;</span>: <span class="string">&#x27;connectivities&#x27;</span>, <span class="string">&#x27;distances_key&#x27;</span>: <span class="string">&#x27;distances&#x27;</span>, <span class="string">&#x27;params&#x27;</span>: &#123;<span class="string">&#x27;n_neighbors&#x27;</span>: <span class="number">10</span>, <span class="string">&#x27;method&#x27;</span>: <span class="string">&#x27;umap&#x27;</span>, <span class="string">&#x27;random_state&#x27;</span>: <span class="number">0</span>, <span class="string">&#x27;metric&#x27;</span>: <span class="string">&#x27;euclidean&#x27;</span>, <span class="string">&#x27;n_pcs&#x27;</span>: <span class="number">10</span>&#125;&#125;), (<span class="string">&#x27;louvain&#x27;</span>, &#123;<span class="string">&#x27;params&#x27;</span>: &#123;<span class="string">&#x27;resolution&#x27;</span>: <span class="number">0.7</span>, <span class="string">&#x27;random_state&#x27;</span>: <span class="number">42</span>&#125;&#125;), (<span class="string">&#x27;paga&#x27;</span>, &#123;<span class="string">&#x27;connectivities&#x27;</span>: &lt;6x6 sparse matrix of <span class="built_in">type</span> <span class="string">&#x27;&lt;class &#x27;</span>numpy.float64<span class="string">&#x27;&gt;&#x27;</span></span><br><span class="line">    <span class="keyword">with</span> <span class="number">14</span> stored elements <span class="keyword">in</span> Compressed Sparse Row <span class="built_in">format</span>&gt;, <span class="string">&#x27;connectivities_tree&#x27;</span>: &lt;6x6 sparse matrix of <span class="built_in">type</span> <span class="string">&#x27;&lt;class &#x27;</span>numpy.float64<span class="string">&#x27;&gt;&#x27;</span></span><br><span class="line">    <span class="keyword">with</span> <span class="number">5</span> stored elements <span class="keyword">in</span> Compressed Sparse Row <span class="built_in">format</span>&gt;, <span class="string">&#x27;groups&#x27;</span>: <span class="string">&#x27;louvain&#x27;</span>, <span class="string">&#x27;pos&#x27;</span>: array([[  <span class="number">9.89749122</span>, -<span class="number">11.9940879</span> ],</span><br><span class="line">       [ <span class="number">12.33582657</span>,  -<span class="number">6.49087334</span>],</span><br><span class="line">       [ <span class="number">16.42411062</span>,  -<span class="number">6.62426792</span>],</span><br><span class="line">       [ <span class="number">12.64145549</span>,  -<span class="number">8.17883229</span>],</span><br><span class="line">       [  <span class="number">8.90823793</span>, -<span class="number">14.36047967</span>],</span><br><span class="line">       [  <span class="number">7.95605852</span>, -<span class="number">13.65646286</span>]])&#125;), (<span class="string">&#x27;louvain_sizes&#x27;</span>, array([<span class="number">407</span>, <span class="number">297</span>, <span class="number">188</span>, <span class="number">108</span>, <span class="number">102</span>,  <span class="number">73</span>])), (<span class="string">&#x27;umap&#x27;</span>, &#123;<span class="string">&#x27;params&#x27;</span>: &#123;<span class="string">&#x27;a&#x27;</span>: <span class="number">0.5830300205483709</span>, <span class="string">&#x27;b&#x27;</span>: <span class="number">1.334166992455648</span>&#125;&#125;), (<span class="string">&#x27;leiden&#x27;</span>, &#123;<span class="string">&#x27;params&#x27;</span>: &#123;<span class="string">&#x27;resolution&#x27;</span>: <span class="number">1</span>, <span class="string">&#x27;random_state&#x27;</span>: <span class="number">0</span>, <span class="string">&#x27;n_iterations&#x27;</span>: -<span class="number">1</span>&#125;&#125;), (<span class="string">&#x27;leiden_colors&#x27;</span>, [<span class="string">&#x27;#1f77b4&#x27;</span>, <span class="string">&#x27;#ff7f0e&#x27;</span>, <span class="string">&#x27;#279e68&#x27;</span>, <span class="string">&#x27;#d62728&#x27;</span>, <span class="string">&#x27;#aa40fc&#x27;</span>, <span class="string">&#x27;#8c564b&#x27;</span>, <span class="string">&#x27;#e377c2&#x27;</span>, <span class="string">&#x27;#b5bd61&#x27;</span>, <span class="string">&#x27;#17becf&#x27;</span>, <span class="string">&#x27;#aec7e8&#x27;</span>, <span class="string">&#x27;#ffbb78&#x27;</span>, <span class="string">&#x27;#98df8a&#x27;</span>, <span class="string">&#x27;#ff9896&#x27;</span>, <span class="string">&#x27;#c5b0d5&#x27;</span>, <span class="string">&#x27;#c49c94&#x27;</span>]), (<span class="string">&#x27;louvain_colors&#x27;</span>, [<span class="string">&#x27;#1f77b4&#x27;</span>, <span class="string">&#x27;#ff7f0e&#x27;</span>, <span class="string">&#x27;#279e68&#x27;</span>, <span class="string">&#x27;#d62728&#x27;</span>, <span class="string">&#x27;#aa40fc&#x27;</span>, <span class="string">&#x27;#8c564b&#x27;</span>])])</span><br><span class="line">With overloaded keys:</span><br><span class="line">    [<span class="string">&#x27;neighbors&#x27;</span>].</span><br></pre></td></tr></table></div></figure>

<p> 而adata.var则是这个样子:</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">adata.var</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">     gene_name     gene_id     n_cells     highly_variable     means     dispersions     dispersions_norm     mean     std</span><br><span class="line">TNFRSF4     TNFRSF4     ENSG00000186827<span class="number">.11</span>     <span class="number">59</span>     <span class="literal">True</span>     <span class="number">0.106400</span>     <span class="number">1.206044</span>     <span class="number">2.019484</span>     <span class="number">0.054150</span>     <span class="number">0.252805</span></span><br><span class="line">TNFRSF18     TNFRSF18     ENSG00000186891<span class="number">.14</span>     <span class="number">39</span>     <span class="literal">True</span>     <span class="number">0.068430</span>     <span class="number">1.109991</span>     <span class="number">1.729919</span>     <span class="number">0.035445</span>     <span class="number">0.202188</span></span><br><span class="line">ATAD3B     ATAD3B     ENSG00000160072<span class="number">.20</span>     <span class="number">139</span>     <span class="literal">False</span>     <span class="number">0.133877</span>     <span class="number">0.300382</span>     -<span class="number">0.710786</span>     <span class="number">0.090129</span>     <span class="number">0.260384</span></span><br><span class="line">THAP3     THAP3     ENSG00000041988<span class="number">.15</span>     <span class="number">75</span>     <span class="literal">False</span>     <span class="number">0.078003</span>     <span class="number">0.520083</span>     -<span class="number">0.048459</span>     <span class="number">0.049195</span>     <span class="number">0.203547</span></span><br><span class="line">unknowngene     unknowngene     ENSG00000260179<span class="number">.1</span>     <span class="number">11</span>     <span class="literal">False</span>     <span class="number">0.011523</span>     <span class="number">0.469297</span>     -<span class="number">0.201563</span>     <span class="number">0.007132</span>     <span class="number">0.078769</span></span><br><span class="line"><span class="meta">... </span>    ...     ...     ...     ...     ...     ...     ...     ...     ...</span><br><span class="line">unknowngene-<span class="number">20032</span>     unknowngene     ENSG00000276256<span class="number">.1</span>     <span class="number">7</span>     <span class="literal">False</span>     <span class="number">0.009656</span>     <span class="number">0.626575</span>     <span class="number">0.272579</span>     <span class="number">0.005590</span>     <span class="number">0.074509</span></span><br><span class="line">unknowngene-<span class="number">20035</span>     unknowngene     ENSG00000273748<span class="number">.1</span>     <span class="number">8</span>     <span class="literal">False</span>     <span class="number">0.006963</span>     <span class="number">0.207106</span>     -<span class="number">0.991981</span>     <span class="number">0.004621</span>     <span class="number">0.059163</span></span><br><span class="line">unknowngene-<span class="number">20038</span>     unknowngene     ENSG00000278817<span class="number">.1</span>     <span class="number">15</span>     <span class="literal">True</span>     <span class="number">0.019199</span>     <span class="number">1.074168</span>     <span class="number">1.621924</span>     <span class="number">0.010459</span>     <span class="number">0.103165</span></span><br><span class="line">unknowngene-<span class="number">20043</span>     unknowngene     ENSG00000278384<span class="number">.1</span>     <span class="number">61</span>     <span class="literal">False</span>     <span class="number">0.067741</span>     <span class="number">0.598364</span>     <span class="number">0.187531</span>     <span class="number">0.041752</span>     <span class="number">0.191105</span></span><br><span class="line">unknowngene-<span class="number">20052</span>     unknowngene     ENSG00000271254<span class="number">.6</span>     <span class="number">8</span>     <span class="literal">True</span>     <span class="number">0.011055</span>     <span class="number">1.205207</span>     <span class="number">2.016962</span>     <span class="number">0.005692</span>     <span class="number">0.078602</span></span><br><span class="line"></span><br><span class="line"><span class="number">25834</span> rows × <span class="number">9</span> columns</span><br></pre></td></tr></table></div></figure>

<p>adata.obs则是：</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">adata.obs</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">n_genes     percent_mito     n_counts     louvain     leiden</span><br><span class="line">barcode                     </span><br><span class="line">AAACCCAAGGAGAGTA     <span class="number">3137</span>     <span class="number">0.112912</span>     <span class="number">9131.0</span>     <span class="number">1</span>     <span class="number">4</span></span><br><span class="line">AAACGCTTCAGCCCAG     <span class="number">2497</span>     <span class="number">0.081725</span>     <span class="number">6424.0</span>     <span class="number">2</span>     <span class="number">3</span></span><br><span class="line">AAAGAACAGACGACTG     <span class="number">2059</span>     <span class="number">0.062348</span>     <span class="number">4940.0</span>     <span class="number">4</span>     <span class="number">10</span></span><br><span class="line">AAAGAACCAATGGCAG     <span class="number">1540</span>     <span class="number">0.068987</span>     <span class="number">3218.0</span>     <span class="number">4</span>     <span class="number">10</span></span><br><span class="line">AAAGAACGTCTGCAAT     <span class="number">2484</span>     <span class="number">0.066250</span>     <span class="number">7366.0</span>     <span class="number">0</span>     <span class="number">8</span></span><br><span class="line"><span class="meta">... </span>    ...     ...     ...     ...     ...</span><br><span class="line">TTTCCTCTCTCTTGCG     <span class="number">3834</span>     <span class="number">0.116246</span>     <span class="number">12723.0</span>     <span class="number">3</span>     <span class="number">11</span></span><br><span class="line">TTTGATCTCTTTGGAG     <span class="number">2567</span>     <span class="number">0.086051</span>     <span class="number">8065.0</span>     <span class="number">0</span>     <span class="number">8</span></span><br><span class="line">TTTGGTTAGTAACCTC     <span class="number">2012</span>     <span class="number">0.083509</span>     <span class="number">5233.0</span>     <span class="number">0</span>     <span class="number">0</span></span><br><span class="line">TTTGGTTGTAGAATAC     <span class="number">3365</span>     <span class="number">0.121682</span>     <span class="number">10585.0</span>     <span class="number">1</span>     <span class="number">1</span></span><br><span class="line">TTTGTTGCAATTAGGA     <span class="number">2001</span>     <span class="number">0.085224</span>     <span class="number">5069.0</span>     <span class="number">2</span>     <span class="number">3</span></span><br><span class="line"></span><br><span class="line"><span class="number">1175</span> rows × <span class="number">5</span> columns</span><br></pre></td></tr></table></div></figure>

<p>adata.X则是：</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">adata.X</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">array([[-<span class="number">0.214197</span>  , -<span class="number">0.17530552</span>, -<span class="number">0.34613955</span>, ..., -<span class="number">0.10138597</span>,</span><br><span class="line">        -<span class="number">0.21847485</span>, -<span class="number">0.0724164</span> ],</span><br><span class="line">       [-<span class="number">0.214197</span>  , -<span class="number">0.17530552</span>, -<span class="number">0.34613955</span>, ..., -<span class="number">0.10138597</span>,</span><br><span class="line">        -<span class="number">0.21847485</span>, -<span class="number">0.0724164</span> ],</span><br><span class="line">       [-<span class="number">0.214197</span>  , -<span class="number">0.17530552</span>, -<span class="number">0.34613955</span>, ..., -<span class="number">0.10138597</span>,</span><br><span class="line">        -<span class="number">0.21847485</span>, -<span class="number">0.0724164</span> ],</span><br><span class="line">       ...,</span><br><span class="line">       [-<span class="number">0.214197</span>  , -<span class="number">0.17530552</span>, -<span class="number">0.34613955</span>, ..., -<span class="number">0.10138597</span>,</span><br><span class="line">        -<span class="number">0.21847485</span>, -<span class="number">0.0724164</span> ],</span><br><span class="line">       [-<span class="number">0.214197</span>  , -<span class="number">0.17530552</span>,  <span class="number">2.2082584</span> , ..., -<span class="number">0.10138597</span>,</span><br><span class="line">        -<span class="number">0.21847485</span>, -<span class="number">0.0724164</span> ],</span><br><span class="line">       [-<span class="number">0.214197</span>  , -<span class="number">0.17530552</span>, -<span class="number">0.34613955</span>, ..., -<span class="number">0.10138597</span>,</span><br><span class="line">        -<span class="number">0.21847485</span>, -<span class="number">0.0724164</span> ]], dtype=float32)</span><br></pre></td></tr></table></div></figure>

<p>从形式也能看到，var和obs是用DataFrame对象存储的，因此，各种pandas的方法都能直接在上面应用，进行一些统计，而X则是np的array，对其进行操作时则需要用numpy的操作方式。而其他非结构化的数据更多的是代码运行中需要的，对研究来说没有特殊意义。此外，知道了这个之后，我们其实就能够手动向var和obs里面添加信息，比如添加gene的type，染色体的正负链，亦或者是样本的临床信息，病理信息，这样的组合型对象非常灵活，对后续分析帮助很大。</p>
<p>下面继续QC的过程，首先，由于这算是比对后的数据，因此需要先在UMI，barcode级别进行一下质控，本质类似于cellranger跑出来的那个html文件，当然，上面的bustools的部分也完成了一点点质控的部分</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">fig, ax = plt.subplots(figsize=(<span class="number">7</span>, <span class="number">7</span>))</span><br><span class="line"></span><br><span class="line">x = np.asarray(adata.X.<span class="built_in">sum</span>(axis=<span class="number">1</span>))[:,<span class="number">0</span>]</span><br><span class="line">y = np.asarray(np.<span class="built_in">sum</span>(adata.X&gt;<span class="number">0</span>, axis=<span class="number">1</span>))[:,<span class="number">0</span>]</span><br><span class="line"></span><br><span class="line">ax.scatter(x, y, color=<span class="string">&quot;green&quot;</span>, alpha=<span class="number">0.25</span>)</span><br><span class="line">ax.set_xlabel(<span class="string">&quot;UMI Counts&quot;</span>)</span><br><span class="line">ax.set_ylabel(<span class="string">&quot;Genes Detected&quot;</span>)</span><br><span class="line">ax.set_xscale(<span class="string">&#x27;log&#x27;</span>)</span><br><span class="line">ax.set_yscale(<span class="string">&#x27;log&#x27;</span>)</span><br><span class="line"></span><br><span class="line">ax.set_xlim(<span class="number">1</span>)</span><br><span class="line">ax.set_ylim(<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">plt.show()</span><br></pre></td></tr></table></div></figure>

<p>这里会出来一个这个“Create a plot showing genes detected as a function of UMI counts.”就是一张横轴为UMI数量，纵轴为基因数量的散点图，如果这张图的趋势基本趋于缓和，说明基本覆盖度已经达到饱和了，在增加测序深度也不会有更多的基因覆盖，当然在1K的PBMC数据里，其实数据基本分布于对角线，缓和趋势不是特别明显，示例数据嘛不强求。</p>
<p>然后就是根据UMI的分布估计细胞数目，这个图应该是最为常见的：</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">expected_num_cells =  <span class="number">1200</span> <span class="comment">#@param &#123;type:&quot;integer&quot;&#125;</span></span><br><span class="line">knee = np.sort((np.array(adata.X.<span class="built_in">sum</span>(axis=<span class="number">1</span>))).flatten())[::-<span class="number">1</span>]</span><br><span class="line"></span><br><span class="line">fig, ax = plt.subplots(figsize=(<span class="number">10</span>, <span class="number">7</span>))</span><br><span class="line"></span><br><span class="line">ax.loglog(knee, <span class="built_in">range</span>(<span class="built_in">len</span>(knee)), linewidth=<span class="number">5</span>, color=<span class="string">&quot;g&quot;</span>)</span><br><span class="line">ax.axvline(x=knee[expected_num_cells], linewidth=<span class="number">3</span>, color=<span class="string">&quot;k&quot;</span>)</span><br><span class="line">ax.axhline(y=expected_num_cells, linewidth=<span class="number">3</span>, color=<span class="string">&quot;k&quot;</span>)</span><br><span class="line"></span><br><span class="line">ax.set_xlabel(<span class="string">&quot;UMI Counts&quot;</span>)</span><br><span class="line">ax.set_ylabel(<span class="string">&quot;Set of Barcodes&quot;</span>)</span><br><span class="line"></span><br><span class="line">plt.grid(<span class="literal">True</span>, which=<span class="string">&quot;both&quot;</span>)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></div></figure>

<p>其实这个图可以和上面bustools inspect的结果里面的预估唯一barcode数进行对照来看，如果偏移得很离谱说明数据有问题，可能是index有问题，或者barcode文件有问题，或者就是测序质量特别差导致UMI分布不均匀。不过一般问题不大。</p>
<p>然后需要看看UMI对应最多的基因是什么（等同于表达量最高的基因）</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sc.pl.highest_expr_genes(adata, n_top=<span class="number">20</span>)</span><br></pre></td></tr></table></div></figure>

<p>这里调用的是scanpy的计算高表达基因的功能，本质上就是计算基因在UMI上的分配情况，可以直接理解为基因的表达量。这里官方给了说明，通常，最容易出现高表达的就是MALAT1，MT基因，RP基因，actin基因，血红蛋白基因，假基因或者ERCC，这是正常的现象。但是当这些基因类型表达高得离谱得时候，比如top30全是某一类基因的时候，就分别与几种异常情况对应了：其中MT基因特别多，说明太多代表细胞存在死亡或者细胞质泄漏的问题，如果ERCC较多说明建库时ERCC加多了技术误差就没法矫正，假基因太多说明比对存在问题或者基因组版本用错了。对于这些基因是否需要先彻底去除后进行分析，目前没有绝对的定论，可以通常是针对<strong>细胞</strong>进行去除，但是对这些基因本身不会直接去掉。</p>
<p>在上面对UMI和barcode的质控没有问题后，进入常见单细胞最熟悉的环节，开始对细胞和基因进行操作，准备进行下游分析了。</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Removes cells with less than 1070 umi counts</span></span><br><span class="line">adata = adata[np.asarray(adata.X.<span class="built_in">sum</span>(axis=<span class="number">1</span>)).reshape(-<span class="number">1</span>) &gt; <span class="number">1070</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># Removes genes with 0 umi counts（也就是有基因，但是没有转录本）</span></span><br><span class="line">adata = adata[:, np.asarray(adata.X.<span class="built_in">sum</span>(axis=<span class="number">0</span>)).reshape(-<span class="number">1</span>) &gt; <span class="number">0</span>]</span><br></pre></td></tr></table></div></figure>

<p>原文这里是说是根据阈值进行过滤，它说是过滤低于1170umi count的细胞，代码是中关于这个axis=i，我看到的最经典的解释是：下标的变化方向；axis=0就是下标沿着第0轴递归操作，axis=1就是下标沿着第1轴递归操作。</p>
<p>第二个操作的目的很容易懂，就是将UMI为0的基因去除，等同于去掉没有测到的基因。但是第一个操作很有意思，它去掉了低于UMI=1070的细胞，但是这个阈值是从哪里来的呢。<br>顺便注意，经过下面两个操作，现在adata本身已经被改变了。因此，可以先将原本的矩阵进行一次保存到一个anndata内置的raw属性里面，这样就能在改动矩阵数据的同时保留原本的表达数据。</p>
<p>接下来就开始正式进入“传统”的表达矩阵类似级别的分析了</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sc.pp.filter_cells(adata, min_genes=<span class="number">200</span>)</span><br><span class="line">sc.pp.filter_genes(adata, min_cells=<span class="number">3</span>)</span><br></pre></td></tr></table></div></figure>

<p>这里就跟seurat非常像了，先筛选细胞里表达不到200个基因的，再筛选在小于3个细胞中表达的基因说白了就是一个是筛低质量细胞，一个筛低质量基因。</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">mito_genes = adata.var_names.<span class="built_in">str</span>.startswith(<span class="string">&#x27;MT-&#x27;</span>)</span><br><span class="line">adata.obs[<span class="string">&#x27;percent_mito&#x27;</span>] = np.<span class="built_in">sum</span>(adata[:, mito_genes].X, axis=<span class="number">1</span>).A1 / np.<span class="built_in">sum</span>(adata.X, axis=<span class="number">1</span>).A1</span><br><span class="line">adata.obs[<span class="string">&#x27;n_counts&#x27;</span>] = adata.X.<span class="built_in">sum</span>(axis=<span class="number">1</span>).A1</span><br><span class="line">sc.pl.violin(adata, [<span class="string">&#x27;n_genes&#x27;</span>, <span class="string">&#x27;n_counts&#x27;</span>, <span class="string">&#x27;percent_mito&#x27;</span>],jitter=<span class="number">0.4</span>, multi_panel=<span class="literal">True</span>)</span><br></pre></td></tr></table></div></figure>

<p>查看质量绘图，对应seurat里的对应的图。<br>n_gene代表细胞内的基因数，可以看到基本在2000-4000（其实貌似也就一般……）<br>nCount可以理解为测序深度（文库大小），后续跟文库normalization有关。<br>percent.mt就是线粒体基因含量。</p>
<p>再根据上面的数据进行一次过滤</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">mask = np.logical_or((adata.obs.n_genes &lt; <span class="number">6500</span>).values, (adata.obs.n_genes &gt; <span class="number">200</span>).values, (adata.obs.percent_mito &lt; <span class="number">0.2</span>).values)</span><br><span class="line">adata = adata[mask, :]</span><br></pre></td></tr></table></div></figure>

<p>然后进行文库归一化，之前我们看到了，测序深度这个问题，并不是在所有细胞中都是一样的（也就是reads不是一样多，基因覆盖度也不一样）。这种情况导致根据UMI弄到的基因并不可比，所以需要矫正，对应的就是Normalize counts，把文库都归一化道10000（也就是seurat里的sizefactor）</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># normalize counts in each cell to be equal</span></span><br><span class="line">sc.pp.normalize_total(adata, target_sum=<span class="number">10</span>**<span class="number">4</span>)</span><br></pre></td></tr></table></div></figure>

<p>然后取表达值的log压缩表达量的范围</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sc.pp.log1p(adata)</span><br></pre></td></tr></table></div></figure>

<p>再来选择高变基因，跟Seurat里一样</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">sc.pp.highly_variable_genes(adata, min_mean=<span class="number">0.01</span>, max_mean=<span class="number">8</span>, min_disp=<span class="number">1</span>, n_top_genes=<span class="number">2000</span>, flavor=<span class="string">&quot;seurat&quot;</span>, n_bins=<span class="number">20</span>)</span><br><span class="line">sc.pl.highly_variable_genes(adata)</span><br><span class="line"></span><br><span class="line"><span class="comment">#此外，虽说提供了“消除”指定“因素”影响的功能，类似于seurat里消除细胞周期的影响。</span></span><br><span class="line"><span class="comment"># sc.pp.regress_out(adata, [&#x27;n_counts&#x27;, &#x27;percent_mito&#x27;])</span></span><br></pre></td></tr></table></div></figure>

<p>然后进行scale:</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sc.pp.scale(adata, max_value=<span class="number">10</span>)</span><br></pre></td></tr></table></div></figure>

<p>其实就是将数据变为0位均值，10为方差的正态分布，才能进行pca。所以为啥要选方差为10…..</p>

        <h2 id="Node5-DownStream-Clustering"   >
          <a href="#Node5-DownStream-Clustering" class="heading-link"><i class="fas fa-link"></i></a><a href="#Node5-DownStream-Clustering" class="headerlink" title="Node5.DownStream-Clustering"></a>Node5.DownStream-Clustering</h2>
      <p>一切准备就绪，上游QC，数据变换都执行完了，准备开始进入聚类分群了</p>
<p>首先是PCA</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># We perform PCA on just the highly variable genes</span></span><br><span class="line">sc.tl.pca(adata, svd_solver=<span class="string">&#x27;arpack&#x27;</span>, use_highly_variable=<span class="literal">True</span>)</span><br><span class="line">sc.pl.pca(adata, color=<span class="string">&#x27;CST3&#x27;</span>)</span><br><span class="line">sc.pl.pca_variance_ratio(adata, log=<span class="literal">True</span>)</span><br><span class="line">sc.pp.neighbors(adata, n_neighbors=<span class="number">20</span>, n_pcs=<span class="number">10</span>)</span><br></pre></td></tr></table></div></figure>

<p>这里的参数也就svd的方法可以改动，其他基本没什么可以改的，注意，如果在之前实际上能看到和示例并不完全相同，应该就是前面flavor的问题，但其实这里效果也挺好的，能看到分得非常开就是了。</p>
<p>此外，scanpy官方教程都说了</p>
<p><em>often, a rough estimate of the number of PCs does fine.</em></p>
<p>然后就可以执行Cluster了，这里有两个路径，其一是：</p>
<p>louvain+paga+umap；</p>
<figure class="highlight python"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">sc.tl.louvain(adata)</span><br><span class="line">sc.tl.paga(adata)</span><br><span class="line">sc.pl.paga(adata, plot=<span class="literal">False</span>)</span><br><span class="line">sc.tl.umap(adata, init_pos=<span class="string">&#x27;paga&#x27;</span>)</span><br></pre></td></tr></table></div></figure>

<p>第二则是：</p>
<p>ledian+umap</p>
<figure class="highlight plaintext"><div class="table-container"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sc.tl.leiden(adata)</span><br><span class="line">sc.pl.umap(adata, color=[&#x27;leiden&#x27;,&#x27;CST3&#x27;, &#x27;NKG7&#x27;, &#x27;PPBP&#x27;])</span><br></pre></td></tr></table></div></figure>

<p>理论上，louvain和ledian不会有太大的区别，一般二者聚类的结果都是基本一致的。</p>
<p>然后就是挑选marker基因进行注释了。这一步这里不讲，这一次主要是上游区别与cellranger的一个流程汇总，注释放在可视化的部分里面专门讲，大概是下下次，下一次就就做一下cellranger的流程，二者用同一份数据对比看看有什么区别。</p>
</div><footer class="post-footer"><div class="post-ending ending"><div class="ending__text">------ 本文结束，感谢您的阅读 ------</div></div><div class="post-copyright copyright"><div class="copyright-author"><span class="copyright-author__name">本文作者: </span><span class="copyright-author__value"><a href="https://biobrick.github.io">Chong Yin</a></span></div><div class="copyright-link"><span class="copyright-link__name">本文链接: </span><span class="copyright-link__value"><a href="https://biobrick.github.io/2021/12/04/%E5%8D%95%E7%BB%86%E8%83%9Ekallisto%E4%BC%A0%E7%BB%9F%E6%B5%81%E7%A8%8B/">https://biobrick.github.io/2021/12/04/%E5%8D%95%E7%BB%86%E8%83%9Ekallisto%E4%BC%A0%E7%BB%9F%E6%B5%81%E7%A8%8B/</a></span></div><div class="copyright-notice"><span class="copyright-notice__name">版权声明: </span><span class="copyright-notice__value">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en" rel="external nofollow" target="_blank">BY-NC-SA</a> 许可协议。转载请注明出处！</span></div></div><nav class="post-paginator paginator"><div class="paginator-prev"><a class="paginator-prev__link" href="/2021/12/04/KM%E7%94%9F%E5%AD%98%E5%88%86%E6%9E%90/"><span class="paginator-prev__icon"><i class="fas fa-angle-left"></i></span><span class="paginator-prev__text">KM生存分析</span></a></div></nav></footer></div></div></div><div class="sidebar-wrap" id="sidebar-wrap"><aside class="sidebar" id="sidebar"><div class="sidebar-nav"><span class="sidebar-nav-toc current">文章目录</span><span class="sidebar-nav-ov">站点概览</span></div><section class="sidebar-toc"><ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#kallisto-bustool-scanpy%E6%B5%81%E7%A8%8B"><span class="toc-number">1.</span> <span class="toc-text">
          kallisto&#x2F;bustool&#x2F;scanpy流程</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Node1-data"><span class="toc-number">2.</span> <span class="toc-text">
          Node1.data</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Node2-Ref"><span class="toc-number">3.</span> <span class="toc-text">
          Node2.Ref</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Node3-count"><span class="toc-number">4.</span> <span class="toc-text">
          Node3.count</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Node4-DownStream-Data-Preprocess"><span class="toc-number">5.</span> <span class="toc-text">
          Node4.DownStream-Data Preprocess</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Node5-DownStream-Clustering"><span class="toc-number">6.</span> <span class="toc-text">
          Node5.DownStream-Clustering</span></a></li></ol></section><!-- ov = overview--><section class="sidebar-ov hide"><div class="sidebar-ov-social"><a class="sidebar-ov-social-item" href="https://github.com/BIOBRICK/" target="_blank" rel="noopener" data-popover="Github" data-popover-pos="up"><span class="sidebar-ov-social-item__icon"><i class="fab fa-github"></i></span></a></div><div class="sidebar-ov-state"><a class="sidebar-ov-state-item sidebar-ov-state-item--posts" href="/archives/"><div class="sidebar-ov-state-item__count">2</div><div class="sidebar-ov-state-item__name">归档</div></a></div><div class="sidebar-ov-cc"><a href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en" target="_blank" rel="noopener" data-popover="知识共享许可协议" data-popover-pos="up"><img src="/images/cc-by-nc-sa.svg"></a></div></section><div class="sidebar-reading"><div class="sidebar-reading-info"><span class="sidebar-reading-info__text">你已阅读了 </span><span class="sidebar-reading-info__num">0</span><span class="sidebar-reading-info__perc">%</span></div><div class="sidebar-reading-line"></div></div></aside></div><div class="clearfix"></div></div></main><footer class="footer" id="footer"><div class="footer-inner"><div><span>Copyright © 2021</span><span class="footer__icon"><i class="fas fa-heart"></i></span><span>Chong Yin</span></div><div><span>主题 - <a href="https://github.com/liuyib/hexo-theme-stun/" title="Stun" target="_blank" rel="noopener">Stun</a></span><span> v2.6.2</span></div></div></footer><div class="loading-bar" id="loading-bar"><div class="loading-bar__progress"></div></div><div class="back2top" id="back2top"><span class="back2top__icon"><i class="fas fa-rocket"></i></span></div></div><script src="https://cdn.jsdelivr.net/npm/jquery@v3.4.1/dist/jquery.min.js"></script><script src="https://cdn.jsdelivr.net/npm/velocity-animate@1.5.2/velocity.min.js"></script><script src="https://cdn.jsdelivr.net/npm/velocity-animate@1.5.2/velocity.ui.min.js"></script><script src="/js/utils.js?v=2.6.2"></script><script src="/js/stun-boot.js?v=2.6.2"></script><script src="/js/scroll.js?v=2.6.2"></script><script src="/js/header.js?v=2.6.2"></script><script src="/js/sidebar.js?v=2.6.2"></script></body></html>